### PR TITLE
Improved cli argument handling

### PR DESCRIPTION
**Pull request by IBims1NicerTobi**

### CODE REVIEW

**Review by VonTum — CHANGES_REQUESTED** [review]
All in all I'm impressed with this change! You seem to have a strong grasp on the way clap 4 does things. Even a few test cases nicely round out this PR. 

I've listed a few issues that still exist, but with those addressed I would happily merge.

File: src/config.rs [hunk 1]
```diff
@@ -26,73 +32,71 @@ pub struct ConfigStruct {
...
-            .help("Generate standalone code with all dependencies in one file of the module specified. "))
+            .help("Generate standalone code with all dependencies in one file of the module specified."))
         .arg(Arg::new("upto")
             .long("upto")
             .help("Describes at what point in the compilation process we should exit early. This is mainly to aid in debugging, where incorrect results from flattening/typechecking may lead to errors, which we still wish to see in say the LSP")
-            .takes_value(true)
-            .possible_values(&["initialize", "flatten", "typecheck", "lint", "instantiate", "codegen"])
-            .default_value("codegen"))
+            .value_parser(clap::builder::EnumValueParser::<EarlyExitUpTo>::new())
+            .default_value("code-gen"))
```
[the remark this answers is not in the document]
Comment (IBims1NicerTobi) [reply src/config.rs:91]: I can manually implement the Trait if you want me to? Then the names would remain backward compatible.

File: src/config.rs [hunk 2]
```diff
@@ -1,21 +1,27 @@
-use std::{cell::UnsafeCell, collections::HashSet, env, ffi::OsStr, path::PathBuf};
-
-use clap::{Arg, Command};
+use clap::{Arg, Command, ValueEnum};
+use std::{
+    collections::HashSet,
+    env,
+    ffi::{OsStr, OsString},
+    path::PathBuf,
+    sync::LazyLock,
```
[the remark this answers is not in the document]
Comment (IBims1NicerTobi) [reply src/config.rs:7]: Fixed

File: src/config.rs [hunk 3]
```diff
@@ -101,86 +105,102 @@ pub fn parse_args() -> Vec<PathBuf> {
...
-        "typecheck" => EarlyExitUpTo::AbstractTypecheck,
-        "lint" => EarlyExitUpTo::Lint,
-        "instantiate" => EarlyExitUpTo::Instantiate,
-        "codegen" => EarlyExitUpTo::CodeGen,
-        _ => unreachable!()
+fn parse_args<I, T>(itr: I) -> Result<ConfigStruct, clap::Error>
+where
+    I: IntoIterator<Item = T>,
+    T: Into<OsString> + Clone,
+{
```
[the remark this answers is not in the document]
Comment (IBims1NicerTobi) [reply src/config.rs:117]: The function signature matches the one from `clap_builder::builder::command::Command::try_get_matches_from` which is used in the in the tests where the passed value is a `&[&'static str]`

[the remark this answers is not in the document]
Comment (VonTum) [reply src/config.rs:117]: Ah I see, then I'll resolve this

File: src/config.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (VonTum) [reply src/config.rs:91]: Ach it's fine, perhaps for debug symbols it's best that the code & flags match. i'll mark it as resolved and merge

**Pull request merged**